compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -4473,7 +4473,7 @@ private void handleForeachDefinitionVariables(VariableDefinitionNode variableDef
         // If the type node is available, we get the type from it.
         BType typeNodeType = symResolver.resolveTypeNode(variableNode.typeNode, blockEnv);
         // Checking whether the RHS type is assignable to LHS type.
-        if (types.isAssignable(varType, typeNodeType)) {
+        if (types.constituentTypesAssignable(varType, typeNodeType)) {
```

Comment:
Why are we changing these? isAssignable should produce correct results.

Reply:
This is a special case. With the modification to `isAssignable` function in this PR, variable with static type of `xml` is not  assignable to `xml<xml:Element>|xml<xml:Comment>|xml<xml:ProcessingInstruction>|xml:Text` or `xml:Element|xml:Comment|xml:ProcessingInstruction|xml:Text`(Which is the correct behaviour and which is expected). 

The function given here is used to validate the types in a foreach loop. In a foreach loop context,  variable with static type `xml` is assignable to loop variable of type `xml<xml:Element>|xml<xml:Comment>|xml<xml:ProcessingInstruction>|xml:Text`. Thus here it is not possible to use `isAssignable`

Example code,

```ballerina
function foo() {
    xml x = xml `<a></a><!--comment-->text<?data?>`;
    foreach xml:Element|xml:Comment|xml:ProcessingInstruction|xml:Text item in x {
        
    }
}
```

Reply:
> The function given here is used to validate the types in a foreach loop. In a foreach loop context, variable with static type `xml` is assignable to loop variable of type `xml<xml:Element>|xml<xml:Comment>|xml<xml:ProcessingInstruction>|xml:Text`. Thus here it is not possible to use `isAssignable`

This definition isn't entirely correct, is it? In the loop/foreach case we should be checking the sequence member type (`xml:Element|xml:Comment|xml:ProcessingInstruction|xml:Text`) against the foreach variable type, not xml.

Reply:
Right, yes.  But we cannot `isAssignable` for this check right, since for variable with static type `xml` can't be assigned to `xml:Element|xml:Comment|xml:ProcessingInstruction|xml:Text`.

Reply:
What I meant was that we should be using `isAssignable` with 

- source type - `xml:Element|xml:Comment|xml:ProcessingInstruction|xml:Text` against
- target type - `xml<xml:Element>|xml<xml:Comment>|xml<xml:ProcessingInstruction>|xml:Text`

which will (should) evaluate to true.

`xml` is `xml<xml:Element|xml:Comment|xml:ProcessingInstruction|xml:Text>`, therefore, iteration value type is `xml:Element|xml:Comment|xml:ProcessingInstruction|xml:Text`. https://ballerina.io/spec/lang/master/#section_5.1.6

Reply:
In the case of type parameter the assignability is checked using the `isAssignable(sourceParam, targetParam)` 
In the case of xml we cannot use this check.
Consider the following code

```ballerina
xml x = xml `<a>a</a>`;
x.forEach(function (xml entry) {
});
```

Here the targetType is `xml:Element|xml:Comment|xml:ProcessingInstruction|xml:Text` and the source type is `xml`. Since an xml value cannot be assigned to ``xml:Element|xml:Comment|xml:ProcessingInstruction|xml:Text`   the check for xml is made special case for type params.